File: app/src/main/java/com/sunny/youyun/wifidirect/manager/WifiDirectManager.java
package com.sunny.youyun.wifidirect.manager;

import android.content.Context;
import android.net.wifi.p2p.WifiP2pConfig;
import android.net.wifi.p2p.WifiP2pManager;
import android.support.annotation.NonNull;

import com.sunny.youyun.wifidirect.client.ClientSocketManager;
import com.sunny.youyun.wifidirect.client.ForwardClientStrategy;
import com.sunny.youyun.wifidirect.config.SocketConfig;
import com.sunny.youyun.wifidirect.model.DeviceInfo;
import com.sunny.youyun.wifidirect.model.TransLocalFile;
import com.sunny.youyun.wifidirect.server.ServerSocketStrategy;

import java.io.IOException;
import java.util.List;


/**
 * Created by qjm3662 on 2016/10/30 0030.
 */

public enum WifiDirectManager implements WifiDirectConnectManager, SocketManager {

    INSTANCE;

    public static WifiDirectManager getInstance() {
        return INSTANCE;
    }

    private WifiDirectConnectManagerImpl wifiDirectConnectManager = null;
    private SocketManagerImpl socketManager = null;
    private boolean isWifiDirectEnable = false;
    private Mode mode = Mode.SINGLE;

    public boolean isWifiDirectEnable() {
        return isWifiDirectEnable;
    }

    public void setWifiDirectEnable(boolean wifiDirectEnable) {
        isWifiDirectEnable = wifiDirectEnable;
    }


    public enum Mode {
        SINGLE, GROUP;
    }

    /**
     * 初始化操作，在Application中作初始化
     *
     * @param c
     */
    public static void init(Context c) {
        System.out.println("init");
        bind(c);
    }

    /**
     * 绑定
     *
     * @param c
     */
    public static void bind(@NonNull Context c) {
        WifiP2pManager manager = (WifiP2pManager) c.getApplicationContext().getSystemService(Context.WIFI_P2P_SERVICE);
        if (manager == null)
            return;
        WifiP2pManager.Channel channel = manager.initialize(c.getApplicationContext(), c.getApplicationContext().getMainLooper(), null);

        INSTANCE.wifiDirectConnectManager = new WifiDirectConnectManagerImpl.Builder()
                .manager(manager)
                .channel(channel)
                .build();

        INSTANCE.socketManager = new SocketManagerImpl.Builder()
                .context(c)
                .build();
    }


    //-------------------------------------回调相关操作--------------------------------------//

//    public ReadFileCallback getReadFileCallback() {
//        return readFileCallback;
//    }
//
//    public SendFileCallBack getSendFileCallBack() {
//        return sendFileCallBack;
//    }
//
//    /**
//     * 绑定文件接收回调
//     * @param readFileCallback
//     */
//    public static void bindReadFileCallback(ReadFileCallback readFileCallback){
//        getInstance().readFileCallback = readFileCallback;
//    }
//
//    /**
//     * 接收文件回调
//     * @param sendFileCallBack
//     */
//    public static void bindSendFileCallBack(SendFileCallBack sendFileCallBack){
//        getInstance().sendFileCallBack = sendFileCallBack;
//    }
//
//    /**
//     * 取消绑定接收文件回调
//     */
//    public static void unBindReadFileCallback(){
//        getInstance().readFileCallback = null;
//    }
//
//    public static void unBindSendFileCallback(){ getInstance().sendFileCallBack = null; }

    //--------------------------------------Socket相关操作------------------------------------------//

    /**
     * 开始Socket监听
     */
    @Override
    public void startServer(@NonNull ServerSocketStrategy strategy) {
        socketManager.startServer(strategy);

    }


    @Override
    public void sendFile(String path, String IP, @NonNull final List<TransLocalFile> mList) {
        socketManager.sendFile(path, IP, mList);
    }

    /**
     * 停止Socket监听
     *
     * @throws IOException Socket关闭异常时抛出
     */
    @Override
    public void stopServer() throws IOException {
        socketManager.stopServer();
    }

    @Override
    public void beginDeviceInfoChange(int port, DeviceInfo deviceInfo) throws IOException {
        socketManager.beginDeviceInfoChange(port, deviceInfo);
    }

    @Override
    public void stopDeviceInfoChange() {
        socketManager.stopDeviceInfoChange();
    }

    @Override
    public void beginReceiveDeviceInfo(int port) {
        socketManager.beginReceiveDeviceInfo(port);
    }

    @Override
    public void beginSendDeviceInfo(int port, DeviceInfo deviceInfo, ForwardClientStrategy strategy) throws IOException {
        socketManager.beginSendDeviceInfo(port, deviceInfo, strategy);
    }

    @Override
    public void beginSendDeviceInfo(int port, DeviceInfo deviceInfo) throws IOException {
        socketManager.beginSendDeviceInfo(port, deviceInfo);
    }

    @Override
    public void stopReceiveDeviceInfo() {
        socketManager.stopReceiveDeviceInfo();
    }

    @Override
    public void stopSendDeviceInfo() {
        socketManager.stopSendDeviceInfo();
    }

    @Override
    public void beginDeviceInfoChange(int port, DeviceInfo deviceInfo, ForwardClientStrategy strategy) throws IOException {
        socketManager.beginDeviceInfoChange(port, deviceInfo, strategy);
    }

    @Override
    public void askIp(String IP) throws IOException {
        ClientSocketManager.getInstance().askIP(IP, SocketConfig.FileListenPort);
    }


    @Override
    public void createGroup(WifiP2pManager.ActionListener createGroupListener) {
        wifiDirectConnectManager.createGroup(createGroupListener);
    }

    @Override
    public void disConnect(WifiP2pManager.ActionListener removeGroupListener) {
        wifiDirectConnectManager.disConnect(removeGroupListener);
    }

    @Override
    public void disConnect() {
        wifiDirectConnectManager.disConnect();
    }

    @Override
    public void cancelConnect(WifiP2pManager.ActionListener actionListener) {
        wifiDirectConnectManager.cancelConnect(actionListener);
    }

    @Override
    public void discover(WifiP2pManager.ActionListener discoverListener) {
        wifiDirectConnectManager.discover(discoverListener);
    }

    @Override
    public void discover() {
        wifiDirectConnectManager.discover();
    }

    @Override
    public void connect(WifiP2pConfig config, WifiP2pManager.ActionListener listener) {
        wifiDirectConnectManager.connect(config, listener);
    }

    @Override
    public void connect(WifiP2pConfig config) {
        wifiDirectConnectManager.connect(config);
    }

    @Override
    public void requestPeers(WifiP2pManager.PeerListListener peerListListener) {
        if (peerListListener == null)
            peerListListener = WifiDirectListeners.getInstance().getPeerListListener();
        wifiDirectConnectManager.requestPeers(peerListListener);
    }

    @Override
    public void requestConnectionInfo(WifiP2pManager.ConnectionInfoListener connectionInfoListener) {
        if (connectionInfoListener == null)
            connectionInfoListener = WifiDirectListeners.getInstance().getConnectionInfoListener();
        wifiDirectConnectManager.requestConnectionInfo(connectionInfoListener);
    }

    @Override
    public void requestGroupInfo(WifiP2pManager.GroupInfoListener groupInfoListener) {
        if (groupInfoListener == null)
            groupInfoListener = WifiDirectListeners.getInstance().getGroupInfoListener();
        wifiDirectConnectManager.requestGroupInfo(groupInfoListener);
    }

    public enum WifiDirectListeners {
        INSTANCE;

        public static WifiDirectListeners getInstance() {
            return INSTANCE;
        }

        private WifiP2pManager.ConnectionInfoListener connectionInfoListener;
        private WifiP2pManager.PeerListListener peerListListener;

        //在android4.3之前不能在Receiver中获取WifiP2pGroup对象
        //要通过requestGroupInfo的方式，在此处处理
        private final WifiP2pManager.GroupInfoListener groupInfoListener = group -> {

            if(group == null)
                return;
            if (group.getOwner() != null) {
                //设置组长MAC
                DeviceInfoManager.getInstance().setGroupOwnerMacAddr(group.getOwner().deviceAddress);
            }

            if (group.isGroupOwner()) {
                //设置组长name
                DeviceInfoManager.getInstance().setGroupOwnerName(group.getNetworkName());

                DeviceInfo deviceInfo = DeviceInfoManager.getInstance().getDeviceInfo();
                //name
                deviceInfo.setDeviceName(group.getNetworkName());
                //macAddr
                deviceInfo.setMacAddress(group.getOwner().deviceAddress);
            } else {
                //设置组长name
                DeviceInfoManager.getInstance().setGroupOwnerName(group.getOwner().deviceName);
            }
        };

        public WifiP2pManager.GroupInfoListener getGroupInfoListener() {
            return groupInfoListener;
        }

        public WifiP2pManager.ConnectionInfoListener getConnectionInfoListener() {
            return connectionInfoListener;
        }

        public void bindConnectionInfoListener(WifiP2pManager.ConnectionInfoListener connectionInfoListener) {
            this.connectionInfoListener = connectionInfoListener;
        }

        public WifiP2pManager.PeerListListener getPeerListListener() {
            return peerListListener;
        }

        public void bindPeerListListener(WifiP2pManager.PeerListListener peerListListener) {
            if (this.peerListListener == peerListListener)
                return;
            this.peerListListener = peerListListener;
        }
    }
}

File: app/src/main/java/com/sunny/youyun/wifidirect/manager/WifiDirectManager.kt
package com.sunny.youyun.wifidirect.manager

import android.content.Context
import android.net.NetworkInfo
import android.net.wifi.WpsInfo
import android.net.wifi.p2p.*
import android.os.Looper
import com.google.gson.Gson
import com.sunny.youyun.wifidirect.model.DeviceInfo
import com.sunny.youyun.wifidirect.socket.client.Client
import com.sunny.youyun.wifidirect.socket.server.Server
import com.sunny.youyun.wifidirect.utils.L
import com.sunny.youyun.wifidirect.utils.doInMainThread
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread
import kotlin.properties.Delegates

/**
 *
 * Created by sunny on 17-11-16.
 */
class WifiDirectManager {

    //Provider by api, use to create p2p connection with other device which support wifi-direct
    private var manager: WifiP2pManager by Delegates.notNull()
    private var channel: WifiP2pManager.Channel by Delegates.notNull()


    var mListener: OnWifiDirectListener? = null
    /**
     * Current Device's Info
     */
    var myDeviceInfo: DeviceInfo? = null

    /**
     * Network Info
     */
    private lateinit var networkInfo: NetworkInfo

    /**
     * the under tow properties use for searching and select one device to connect
     */
    lateinit var wifiP2pDeviceList: WifiP2pDeviceList
    val wifiDeviceList = mutableListOf<DeviceInfo>()
    val groupList = WifiP2pDeviceInfoListManager.INSTANCE

    private lateinit var wifiP2pGroup: WifiP2pGroup

    companion object {
        val INSTANCE = WifiDirectManager()
    }

    fun init(context: Context): WifiDirectManager {
        manager = context.getSystemService(Context.WIFI_P2P_SERVICE) as WifiP2pManager
        channel = manager.initialize(context, Looper.getMainLooper()) {

        }
        myDeviceInfo = DeviceInfo(null, "", null)
        return this
    }


    /////////////////////////////////////////////////////////////////////////////////
    //////////// Provider Some interface to record info about WifiDirect
    /////////////////////////////////////////////////////////////////////////////////
    /**
     * Update Device info
     */
    fun updateDeviceInfo(wifiP2pInfo: WifiP2pInfo): WifiDirectManager {
        myDeviceInfo?.groupOwnerAddr = wifiP2pInfo.groupOwnerAddress
        myDeviceInfo?.isGroupOwner = wifiP2pInfo.isGroupOwner
        if (myDeviceInfo?.isGroupOwner == true)
            myDeviceInfo?.deviceAddress = wifiP2pInfo.groupOwnerAddress
        println("register Device Info: ")
        println(Gson().toJson(wifiP2pInfo))
        //if group formed, and this device is not the group owner
        //then send a simple socket to group owner for register itself
        if (wifiP2pInfo.groupFormed && !wifiP2pInfo.isGroupOwner) {
            L.i("try to register")
            Client.INSTANCE
                    .register(wifiP2pInfo.groupOwnerAddress)
        }

        //if group formed, and this device is the group owner
        //then open a socket to listen the request from client
        if (wifiP2pInfo.groupFormed && wifiP2pInfo.isGroupOwner) {
            Server.INSTANCE
                    .startSync()
        }
        if(myDeviceInfo != null)
            doInMainThread {
                mListener?.onDeviceInfoChange(myDeviceInfo!!)
            }
        return this
    }

    fun updateDeviceInfo(wifiP2pDevice: WifiP2pDevice): WifiDirectManager {
        myDeviceInfo?.mac = wifiP2pDevice.deviceAddress
        myDeviceInfo?.deviceName = wifiP2pDevice.deviceName
        return this
    }

    fun clearDeviceInfo() {
        myDeviceInfo?.isGroupOwner = false
        myDeviceInfo?.groupOwnerAddr = null
        myDeviceInfo?.deviceAddress = null
        myDeviceInfo?.mac = ""
        myDeviceInfo?.deviceName = ""
    }

    /**
     * Update NetWork info
     */
    fun updateNetWorkInfo(networkInfo: NetworkInfo): WifiDirectManager {
        this.networkInfo = networkInfo
        doInMainThread {
            mListener?.onNetWorkInfoChange(networkInfo)
        }
        return this
    }

    /**
     * register the available peer list
     */
    fun updateWifiP2pDeviceList(list: WifiP2pDeviceList): WifiDirectManager {
        doAsync {
            wifiP2pDeviceList = list
            wifiDeviceList.clear()
            list.deviceList
                    .forEach {
                        wifiDeviceList.add(DeviceInfo(null, it.deviceAddress, null,
                                it.isGroupOwner, it.deviceName))
                    }
            uiThread {
                mListener?.onWifiP2pDeviceListChange(wifiDeviceList)
            }
        }
        return this
    }

    fun update(deviceInfo: DeviceInfo): WifiDirectManager {
        groupList.register(deviceInfo)
        doInMainThread {
            mListener?.onWifiP2pDeviceListChange(groupList.mList)
        }
        return this
    }


    /**
     * register WifiP2pGroup info
     * this information normal use in communication
     */
    fun updateWifiP2pGroup(wifiP2pGroup: WifiP2pGroup): WifiDirectManager {
        this.wifiP2pGroup = wifiP2pGroup
        groupList.updateList(wifiP2pGroup.clientList)
        doInMainThread {
            mListener?.onWifiP2pGroupChange(wifiP2pGroup)
        }
        return this
    }


    ////////////////////////////////////////////////////////////////////
    //////////// Provide some interface to easy operate the origin API provider by Google
    ////////////////////////////////////////////////////////////////////
    /**
     * request the peer list available
     */
    fun requestPeers(): WifiDirectManager {
        manager.requestPeers(channel) { peers ->
            updateWifiP2pDeviceList(peers)
        }
        return this
    }

    /**
     * request the group info
     */
    fun requestGroup(): WifiDirectManager {
        manager.requestGroupInfo(channel) { group ->
            updateWifiP2pGroup(group)
        }
        return this
    }

    /**
     * discover available peer list
     */
    fun discoverPeers(): WifiDirectManager {
        manager.discoverPeers(channel, object : WifiP2pManager.ActionListener {
            override fun onSuccess() {
                println("discover success")
                doInMainThread {
                    mListener?.discoverPeers(true)
                }
            }

            override fun onFailure(reason: Int) {
                println("discover fail: $reason")
                doInMainThread {
                    mListener?.discoverPeers(false, reason)
                }
            }

        })
        return this
    }


    /**
     * create a group
     */
    fun createGroup(): WifiDirectManager {
        manager.createGroup(channel, object : WifiP2pManager.ActionListener {
            override fun onSuccess() {
                println("create group success")
                doInMainThread {
                    mListener?.createGroup(true)
                }
            }

            override fun onFailure(reason: Int) {
                println("create group fail: $reason")
                doInMainThread {
                    mListener?.createGroup(false, reason)
                }
            }
        })
        return this
    }

    /**
     * remove a group
     */
    fun removeGroup(): WifiDirectManager {
        manager.removeGroup(channel, object : WifiP2pManager.ActionListener {
            override fun onSuccess() {
                println("remove group success")
                // when remove a group, clear the device info
                INSTANCE
                        .clearDeviceInfo()
                Server.INSTANCE
                        .stop()
                doInMainThread {
                    mListener?.removeGroup(true)
                }
            }

            override fun onFailure(reason: Int) {
                println("remove group success: $reason")
                doInMainThread {
                    mListener?.removeGroup(false, reason)
                }
            }

        })
        return this
    }

    /**
     * connect by MAC address(hardware address)
     */
    fun connect(deviceAddress: String) {
        val config = WifiP2pConfig()
        config.deviceAddress = deviceAddress
        config.wps.setup = WpsInfo.PBC
        manager.connect(channel, config, object : WifiP2pManager.ActionListener {
            override fun onSuccess() {
                println("connect operator success")
                doInMainThread {
                    mListener?.connect(true)
                }
            }

            override fun onFailure(reason: Int) {
                println("connect operator fail: " + reason)
                doInMainThread {
                    mListener?.connect(false, reason)
                }
            }
        })
    }

    /**
     * invoke this method to connect a p2p device
     */
    fun connect(device: WifiP2pDevice): WifiDirectManager {
        connect(device.deviceAddress)
        return this
    }

    fun cancelConnect(): WifiDirectManager {
        manager.cancelConnect(channel, object : WifiP2pManager.ActionListener {
            override fun onSuccess() {
                println("cancel connect success")
                mListener?.onCancelConnect(true)
            }

            override fun onFailure(reason: Int) {
                println("cancel connect fail: $reason")
                mListener?.onCancelConnect(false, reason)
            }
        })
        return this
    }

    /**
     * Provide this method to re_name current device
     * Because this method is hide by API
     * so I try to do this by java reflection
     */
    fun setDeviceName(name: String): WifiDirectManager {
        //利用反射试图修改设备的名字
        val c = WifiP2pManager::class.java
        val m = c.getMethod("setDeviceName", WifiP2pManager.Channel::class.java, String::class.java,
                WifiP2pManager.ActionListener::class.java)
        m.isAccessible = true
        m.invoke(manager, channel, name, object : WifiP2pManager.ActionListener {
            override fun onSuccess() {
                L.i("modify device info success")
            }

            override fun onFailure(reason: Int) {
                L.i("modify device info fail: $reason")
            }

        })
        return this
    }


    /////////////////////////////////////////////////////////////////////////
    /////////// Define callback
    /////////////////////////////////////////////////////////////////////////

    interface OnWifiDirectListener {
        fun createGroup(isSuccess: Boolean, reason: Int = -1) {

        }

        fun removeGroup(isSuccess: Boolean, reason: Int = -1) {

        }

        fun discoverPeers(isSuccess: Boolean, reason: Int = -1) {

        }

        fun onDeviceInfoChange(deviceInfo: DeviceInfo) {

        }

        fun onNetWorkInfoChange(networkInfo: NetworkInfo) {

        }

        fun onWifiP2pDeviceListChange(list: MutableList<DeviceInfo>) {

        }

        fun onWifiP2pGroupChange(wifiP2pGroup: WifiP2pGroup) {

        }

        fun connect(isSuccess: Boolean, reason: Int = -1) {

        }

        fun onCancelConnect(isSuccess: Boolean, reason: Int = -1) {

        }

        /**
         * when the client connect to Group owner and register success
         * this method will be invoked
         */
        fun onRegisterSuccess() {

        }

        fun onNewDeviceRegister(deviceInfo: DeviceInfo) {

        }
    }

}
